src/helpers/similarity.ts

```diff
@@ -0,0 +1,86 @@
+import { getBotContext } from "../bindings";
+import { Payload, Choices } from "../types";
+import { listAllIssuesForRepo, upsertCommentToIssue } from "../helpers/issue";
+import axios from "axios";
+
+export const findDuplicateOne = async () => {
+  const { payload: _payload } = getBotContext();
+  const issue = (_payload as Payload).issue;
+  const issues = await listAllIssuesForRepo("all");
+  issues.shift();
+  const importantWords = await extractImportantWords();
+  const wordCount = importantWords.length;
+  for (const iss of issues) {
+    if (iss.body) {
+      const probability = (countIncludedWords(iss.body, importantWords) * 100.0) / wordCount;
+      if (probability > parseInt(process.env.SIMILARITY_THRESHOLD || "80")) {
+        if (issue?.number) {
+          await upsertCommentToIssue(issue?.number, `Similar issue (${iss.title}) found at ${iss.html_url}.\nSimilarity is about ${probability}%`, "created");
+          return;
+        }
+      }
+    }
+  }
+};
+
+const countIncludedWords = (inputString: string, words: string[]): number => {
+  const lowerCaseString = inputString.toLowerCase(); // Convert the string to lowercase for case-insensitive matching
+
+  let count = 0;
+  for (const word of words) {
+    if (lowerCaseString.includes(word.toLowerCase())) {
+      count++;
+    }
+  }
+
+  return count;
+};
+
+export const extractImportantWords = async (): Promise<string[]> => {
+  const { payload: _payload } = getBotContext();
+  const issue = (_payload as Payload).issue;
+  const { result: res } = await getAnswerFromChatGPT(`'Issue title: "${issue?.title}"\nIssue body: "${issue?.body}"'`);
+  return res.split("#");
+};
+
+export const getAnswerFromChatGPT = async (prompt: string): Promise<{ result: string }> => {
+  const body = JSON.stringify({
+    model: "gpt-3.5-turbo",
+    messages: [
+      {
+        role: "system",
+        content:
+          process.env.CHATGPT_SYSTEM_PROMPT_FOR_IMPORTANT_WORDS ||
+          "You are an 'important words finder'. You need to find important words from given context. You only have to give important words from given context and you have to separate the words by #",
+      },
+      {
+        role: "user",
+        content:
+          (process.env.CHATGPT_USER_PROMPT_FOR_IMPORTANT_WORDS ||
+            "I need your help to find duplicate issues on my GitHub repository. For context, the entire strategy is the following:\n\n1. A new issue is posted\n2. We ask you to extract a word list of the most \"important\" (i.e. unique adjectives?) words.\n3. We search the repository for all issues with the important words.\n4. We go from highest issue number (most recent) and read the issue description.\n5. If >80% confidence that it's a redundant issue, stop the search and link back to it with a warning saying that it's likely to be a duplicate.\nRight now, we are on step 2.\n") +
+          prompt,
+      },
+    ],
+    max_tokens: 1500,
+    temperature: 0,
```

Comment:
Yes, temperature is related to randomness of the answer. I set it to 0 because we don't need random answer...

Reply:
Would it be good to move temperature to botconfig?

Reply:
I don't think we need temperature (randomness) to be customisable. @pavlovcik rfc

Reply:
temperature and similarity threshold can be feature configs and not openai config. They can go in repo config. Eventually, I think openai here will be used for other purposes too. 

`duplicate-issue-ai-temperature`
`duplicate-issue-similarity-threshold`

What do you think @pavlovcik

Reply:
Or feature config stored as repo config
```
duplicate-issue-ai-system-prompt
duplicate-issue-ai-user-prompt
duplicate-issue-ai-temperature
duplicate-issue-similarity-threshold
```
env var
```
OPENAI_API_HOST
OPENAI_API_KEY=
```

Reply:
> Eventually, openai here will be used for other purposes too. The following can be feature-specific config stored as repo config
> 
> ```
> 
> duplicate-issue-ai-system-prompt
> 
> duplicate-issue-ai-user-prompt
> 
> duplicate-issue-ai-temperature
> 
> duplicate-issue-similarity-threshold
> 
> ```
> 
> 
> 
> What do you think @pavlovcik 

I think we can consider specific settings (prompt, temperature) for specific capabilities. Like the ones that it makes sense to customize like this one proposal to give the bot a "personality" 

However for this duplicates one I agree no need to expose properties for customization.

Reply:
> I think we can consider specific settings (prompt, temperature) for specific capabilities. Like the ones that it makes sense to customize like [this one proposal](https://github.com/ubiquity/research/issues/58) to give the bot a "personality"
> 
> (It's in the research repository because I want to discuss the concept on a call with the research team before we make it official.)
> 
> Perhaps it could be useful to expose properties for customization in case repository maintainers want to add additional context to enhance duplicate issue finding accuracy on their specific repositories (e.g. explicitly listing out the "important" words.)

I am a bit confused now about where to store each keys. So should OPENAI_API_KEY and OPENAI_API_HOST be stored in .env or /config/shared.ts ?
Other similarity configs including duplicate-issue-ai-system-prompt, duplicate-issue-ai-user-prompt, duplicate-issue-ai-temperature, duplicate-issue-similarity-threshold
should be stored in repo config?
@pavlovcik RFC

Reply:
@0xcodercrane knows best

Reply:
@0xcodercrane rfc

Reply:
you should store them in `.env`

Reply:
Okay, got it.